### PR TITLE
feat(translator): implement httproutefilter host rewrite

### LINKED ISSUE #3173: Support additional host rewrite options
*Description*:

Envoy supports multiple host header rewriting options:
- Literal rewriting: the host header is rewritten to a statically defined value
- Automatic rewriting: the host header is rewritten with the [DNS name](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/address.proto#envoy-v3-api-msg-config-core-v3-socketaddress) or endpoint [hostname](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#envoy-v3-api-field-config-endpoint-v3-endpoint-hostname) of the selected envoy cluster. 
- Header-based rewriting: envoy rewrites the host header with a value consumed from a different header. The different header name is statically defined in
- Path regex rewriting: the host header is rewritten with a value derived by applying a regex to the request path. 

Unlike other headers, it's safer to rewrite the host header in the forwarding phase, to avoid possible issues related to failed route recalculation after the header is rewritten. Envoy's host rewrite feature are implemented in this way. 

Gateway-API and Envoy gateway support literal [host rewriting](https://gateway-api.sigs.k8s.io/guides/http-redirect-rewrite/?h=rewrite#rewrites). 

Scenarios:
- Automatic rewriting: When multiple backends are configured (e.g. for weighted traffic splitting), it could be necessary to rewrite the hostname automatically, based on the dynamically selected backend. 
- Header-based rewriting: allows envoy to rewrite the host header based on information provided as a hint by the end user or extracted from the request (e.g. a jwt claim with the desired target host header transformed to a header)
- Path-Regex rewriting: rewriting to a subdomain that is encoded as a path parameter in a request to the top-level domain.  

Envoy Gateway can support additional host rewriting options by adding new options to a policy or implementing EG-specific HTTP Route Filters. 

Application developers are the likely users of this feature, as they know the desired host header value and structure in the backend. 

Prior art: these features are supported by some gateways in the K8s space: 
- Nginx Ingress support the annotation [upstream-vhost](https://github.com/kubernetes/ingress-nginx/blob/main/docs/user-guide/nginx-configuration/annotations.md#custom-nginx-upstream-vhost), which can be used to inject dynamic values such as value from other headers, using nginx variables, e.g. `nginx.ingress.kubernetes.io/upstream-vhost: $http_x_custom_header`
- Gloo Edge supports all of envoy's [host rewrite options](https://docs.solo.io/gloo-edge/latest/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/options.proto.sk/): `hostRewrite `, `autoHostRewrite`, `hostRewritePathRegex `, `hostRewriteHeader`
- Contour, Ambassador and Istio support literal host rewriting only.

*Relevant Links*
- https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-auto-host-rewrite
- https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-host-rewrite-header
- https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-host-rewrite-path-regex

### PR DESCRIPTION
**Pull request by guydc**
**What type of PR is this?**

**What this PR does / why we need it**:
- Updates API to format requested in #4410
- Implements host rewrite based on custom header and backend dns name
- E2E and docs
- Fix Regex Path rewrite docs and translation bug

**Which issue(s) this PR fixes**:

Fixes #3173, #4410

### CODE REVIEW

File: internal/gatewayapi/filters.go [hunk 1]
```diff
@@ -152,9 +153,18 @@ func (t *Translator) processURLRewriteFilter(
...
+		// 2. HTTPRouteFilter and URLRewriteFilter must not change the same component (host, path)
+
+		hasCoreRewrites := filterContext.URLRewrite.Path != nil && (filterContext.URLRewrite.Path.FullReplace != nil ||
+			filterContext.URLRewrite.Path.PrefixMatchReplace != nil) || (filterContext.URLRewrite.Host != nil && filterContext.URLRewrite.Host.Name != nil)
+		hasExtensionPathRewrites := filterContext.URLRewrite.Path != nil && filterContext.URLRewrite.Path.RegexMatchReplace != nil
+		hasExtensionHostRewrites := filterContext.URLRewrite.Host != nil && (filterContext.URLRewrite.Host.Header != nil ||
+			filterContext.URLRewrite.Host.Backend != nil)
+
+		if (rewrite.Hostname != nil && (hasCoreRewrites || hasExtensionHostRewrites)) ||
+			(rewrite.Path != nil && (hasCoreRewrites || hasExtensionPathRewrites)) {
```
Comment (zirain) [inline internal/gatewayapi/filters.go:167]: any chace to improve these lines, it's hard to read/understand( `conditionA && (conditionB || conditionC)`).

File: internal/gatewayapi/filters.go [hunk 2]
```diff
@@ -751,48 +763,22 @@ func (t *Translator) processExtensionRefHTTPFilter(extFilter *gwapiv1.LocalObjec
...
+					if filterContext.URLRewrite != nil {
+						// avoid conflicts:
+						// 1. multiple extension HTTPRouteFilter (of any type)
+						// 2. HTTPRouteFilter and URLRewriteFilter must not change the same component (host, path)
+
+						hasCorePathRewrites := filterContext.URLRewrite.Path != nil && (filterContext.URLRewrite.Path.FullReplace != nil ||
+							filterContext.URLRewrite.Path.PrefixMatchReplace != nil)
+						hasCoreHostnameRewrites := filterContext.URLRewrite.Host != nil && filterContext.URLRewrite.Host.Name != nil
+						hasExtensionRewrites := (filterContext.URLRewrite.Path != nil && filterContext.URLRewrite.Path.RegexMatchReplace != nil) ||
+							(filterContext.URLRewrite.Host != nil && (filterContext.URLRewrite.Host.Header != nil || filterContext.URLRewrite.Host.Backend != nil))
```
Comment (zirain) [inline internal/gatewayapi/filters.go:778]: duplicated with the above

File: api/v1alpha1/httproutefilter_types.go [hunk 3]
```diff
@@ -37,11 +37,10 @@ type HTTPRouteFilterSpec struct {
 
 // HTTPURLRewriteFilter define rewrites of HTTP URL components such as path and host
 type HTTPURLRewriteFilter struct {
-	// Hostname is the value to be used to replace the Host header value during
+	// Hostname is the value to be used to replace the Hostname header value during
```
Comment (zhaohuabing) [inline api/v1alpha1/httproutefilter_types.go:40]: Nit: `HostName` Header should be `Host` header.
```suggestion
	// Hostname is the value to be used to replace the Host header value during
```

File: api/v1alpha1/httproutefilter_types.go [hunk 4]
```diff
@@ -63,12 +62,12 @@ const (
 type HTTPHostnameModifierType string
 
 const (
-	// HeaderHTTPHostnameModifier indicates that the Host header value would be replaced with the value of the header specified in setFromHeader.
+	// HeaderHTTPHostnameModifier indicates that the Hostname header value would be replaced with the value of the header specified in header.
```
Comment (zhaohuabing) [inline api/v1alpha1/httproutefilter_types.go:65]: Same here.

File: api/v1alpha1/httproutefilter_types.go [hunk 5]
```diff
@@ -63,12 +62,12 @@ const (
...
 	// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-host-rewrite-header
-	HeaderHTTPHostnameModifier HTTPHostnameModifierType = "SetFromHeader"
-	// BackendHTTPHostnameModifier indicates that the Host header value would be replaced by the DNS name of the backend if it exists.
+	HeaderHTTPHostnameModifier HTTPHostnameModifierType = "Header"
+	// BackendHTTPHostnameModifier indicates that the Hostname header value would be replaced by the DNS name of the backend if it exists.
```
Comment (zhaohuabing) [inline api/v1alpha1/httproutefilter_types.go:68]: Same here.

**Review by zirain — DISMISSED** [review]

**Review by zirain — APPROVED** [review]

**Review by arkodg — APPROVED** [review]

**Pull request merged**